### ISSUE TITLE
Relion - Switching to Single Precision?

### ISSUE THREAD

**Issue by JusticeForMikeBrown** [3dem/relion#309]
Hello

We have been doing our runs in Double precision on CPU and Single on GPU which remains the default.  Recently we considered moving to Single on CPU and Double on GPU since we now have much faster GPUs.

However it sounds like that isn't advisable while we see the informative warning message:

```Running CPU instructions in single precision. Runs might not be exactly reproducible```

We very much need reproducibility so can you advise whether single should be used at this time?

**Comment by JusticeForMikeBrown** [comment]
After perusing [Accelerated cryo-EM structure determination with parallelisation using GPUs in RELION-2](https://elifesciences.org/articles/18722) I am wondering if there are still steps which are precision sensitive thus must still be ran on CPU in double precision.

**Comment by bforsbe** [comment]
The GPU-side of RELION has been engineered to deal with lower precision, because single precision is so much faster there. On the CPU-side, the difference is much smaller, so efforts have not been made to deal with lower precision. Therefore, you currently need to run CPU-instructions in double. You can run it in single, but I'd suggest you use a whole bunch of caution in that case. 

As for running GPU-instructions in double; don't do it. Depending on your GPU, you may have ok performance using double precision, but it will never be better than single precision. Also, the increased precision means very little for the GPU-side calculations. Basically, you would be using double precision to represent an image which is mostly noise. On a per-image basis, double precision is simply excessive. 

In conclusion; 

Good performance and good results: CPU-double & GPU-single

Bad performance and bad results; CPU-single & GPU-double

Good performance but bad results; all-single

Bad performance but good results; all-double.

**Comment by JusticeForMikeBrown** [comment]
Thanks for the extremely informative response!

Could Relion support half precision on GPUs in future or would the numbers not be long enough.

**Comment by bforsbe** [comment]
The precision might be enough. That is an open question which we don't really know the answer to.

As for RELION using it, that depends on the incentive to implement that capacity. The hardware needs to be attractive for it to be a good idea, and currently you need to buy really high-end GPUs to work efficiently with half-precision logics. So there isn't much gain in those developments. From an I/O point of view there might be (healf of the file-size to ship around over a cluster network, e.g.), but then we might as well just read small (on disk) images in a lower precision and promote those numbers to single-precision as soon as we want to use them for calculations. That makes more sense, currently.
